### PR TITLE
Feature: masking padded sequence

### PR DESCRIPTION
**Pull request by ulricharmel**
## What?

- Implement the padded sequence masking on top of the develop branch so that this can easily be merged with other PRs.

## Why?
- Masking padded sequences improves results.
## How?
- Added code in step.py to create a mask from the discounts and apply this mask to all training data variables.
## Extra
[Benchmarks](https://www.notion.so/instadeep/Sequence-padding-and-large-batch-sizes-f70ce4feb33340c3939b5df6114a26e6)

### CODE REVIEW

File: mava/components/training/losses.py [hunk 1]
```diff
@@ -19,22 +19,68 @@
...
+            update.
+    """
+    chex.assert_rank([prob_ratios_t, adv_t], [1, 1])
+    chex.assert_type([prob_ratios_t, adv_t], [float, float])
+
+    adv_t = jax.lax.select(use_stop_gradient, jax.lax.stop_gradient(adv_t), adv_t)
+    clipped_ratios_t = jnp.clip(prob_ratios_t, 1.0 - epsilon, 1.0 + epsilon)
+    clipped_objective = jnp.fmin(prob_ratios_t * adv_t, clipped_ratios_t * adv_t)
+
+    return -jnp.sum(clipped_objective * mask) / jnp.sum(mask)
```
Comment (OmaymaMahjoub) [inline mava/components/training/losses.py:77]: This could lead to an error that happen to us priorly, at some cases you may divide by zero `jnp.sum(mask)=0`(I know it may be rare but at least add a small factor next to it)

File: mava/components/training/losses.py [hunk 2]
```diff
@@ -259,20 +313,20 @@ def policy_loss_fn(
                             policy_params, observations
                         )
 
-                    log_probs = network.log_prob(distribution_params, actions)
-                    entropy = network.entropy(distribution_params)
+                    log_probs = network.log_prob(distribution_params, actions) * mask
+                    entropy = network.entropy(distribution_params) * mask
```
Comment (OmaymaMahjoub) [inline mava/components/training/losses.py:317]: If you can make the masking optional so in case we don't need it we set the flag to false

File: mava/components/training/losses.py [hunk 3]
```diff
@@ -357,11 +416,11 @@ def critic_loss_fn(
...
                             clipped_value_error
                         )
-                        value_loss = jnp.mean(
+                        value_loss = jnp.sum(
                             jnp.fmax(unclipped_value_loss, clipped_value_loss)
-                        )
+                        ) / jnp.sum(mask)
                     else:
-                        value_loss = jnp.mean(unclipped_value_loss)
+                        value_loss = jnp.sum(unclipped_value_loss) / jnp.sum(mask)
```
Comment (OmaymaMahjoub) [inline mava/components/training/losses.py:423]: same as above dividing by zero in rare cases

File: mava/components/training/losses.py [hunk 4]
```diff
@@ -357,11 +416,11 @@ def critic_loss_fn(
                         clipped_value_loss = trainer.store.value_loss_fn(
                             clipped_value_error
                         )
-                        value_loss = jnp.mean(
+                        value_loss = jnp.sum(
                             jnp.fmax(unclipped_value_loss, clipped_value_loss)
-                        )
+                        ) / jnp.sum(mask)
```
Comment (OmaymaMahjoub) [inline mava/components/training/losses.py:421]: same as above dividing by zero in rare cases

File: mava/components/training/losses.py [hunk 5]
```diff
@@ -259,20 +313,20 @@ def policy_loss_fn(
...
 
-                    policy_loss = rlax.clipped_surrogate_pg_loss(
-                        rhos, advantages, clipping_epsilon
+                    policy_loss = clipped_surrogate_pg_loss(
+                        rhos, advantages, clipping_epsilon, mask
                     )
 
                     # Entropy regulariser.
-                    entropy_loss = -jnp.mean(entropy)
+                    entropy_loss = -jnp.sum(entropy) / jnp.sum(mask)
```
Comment (OmaymaMahjoub) [inline mava/components/training/losses.py:329]: same as above dividing by zero in rare cases

**Review by OmaymaMahjoub — CHANGES_REQUESTED** [review]
Thanks @ulricharmel, just few changes to edit

File: mava/components/training/losses.py (hunk 1 above)
Comment (ulricharmel) [reply mava/components/training/losses.py]: Yep. I just pushed some updates for with and also added a flag for the sequence masking.

**Review by OmaymaMahjoub — DISMISSED** [review]
Thanks @ulricharmel, I think it is ready to go :fire:

File: mava/components/training/losses.py (hunk 1 above)
Comment (OmaymaMahjoub) [reply mava/components/training/losses.py]: Cool thanks :slightly_smiling_face:

File: examples/smac/feedforward/decentralised/run_ippo_smac_death_masking.py [hunk 6]
```diff
@@ -73,15 +74,16 @@ def network_factory(*args: Any, **kwargs: Any) -> Any:
...
     )
 
     critic_optimiser = optax.chain(
-        optax.clip_by_global_norm(40.0), optax.scale_by_adam(), optax.scale(-1e-4)
+        optax.clip_by_global_norm(40.0), optax.scale_by_adam(), optax.scale(-5e-4)
     )
 
     # Create the system.
     system = ippo.IPPOSystem()
+    system.update(HuberValueLoss)
```
Comment (DriesSmit) [inline examples/smac/feedforward/decentralised/run_ippo_smac_death_masking.py:86]: Do we need Huber loss when testing the death masking code? Dit is not work with vanilla IPPO?

File: examples/smac/feedforward/decentralised/run_ippo_smac_death_masking.py [hunk 7]
```diff
@@ -92,10 +94,21 @@ def network_factory(*args: Any, **kwargs: Any) -> Any:
         policy_optimiser=policy_optimiser,
         critic_optimiser=critic_optimiser,
         run_evaluator=True,
-        epoch_batch_size=5,
+        epoch_batch_size=32,
         num_epochs=15,
         num_executors=1,
-        multi_process=True,
+        multi_process=False,
```
Comment (DriesSmit) [inline examples/smac/feedforward/decentralised/run_ippo_smac_death_masking.py:100]: Why did we disable multi-process?

File: examples/smac/feedforward/decentralised/run_ippo_smac_death_masking.py [hunk 8]
```diff
@@ -92,10 +94,21 @@ def network_factory(*args: Any, **kwargs: Any) -> Any:
...
+        sequence_length=121,
+        period=121,
```
Comment (DriesSmit) [inline examples/smac/feedforward/decentralised/run_ippo_smac_death_masking.py:102]: Did we decide that this is the only fix to our problem? There was no solution that still allows for smaller sequences. I think the code uses smaller sequences, but the full sequence is used to calculate GAE? Or do they also train on the full sequence?

File: mava/components/training/losses.py [hunk 9]
```diff
@@ -259,20 +315,21 @@ def policy_loss_fn(
                             policy_params, observations
                         )
 
-                    log_probs = network.log_prob(distribution_params, actions)
-                    entropy = network.entropy(distribution_params)
+                    log_probs = network.log_prob(distribution_params, actions) * mask
```
Comment (DriesSmit) [inline mava/components/training/losses.py:318]: Is it necessary to mask the log_probs f we mask the clipped_surrogate_pg_loss? Maybe it is necessary. Just confirm this.

File: mava/components/training/losses.py [hunk 10]
```diff
@@ -332,10 +391,13 @@ def critic_loss_fn(
                     observations: Any,
                     target_values: jnp.ndarray,
                     behavior_values: jnp.ndarray,
+                    mask: jnp.ndarray,
                 ) -> Tuple[jnp.ndarray, Dict[str, jnp.ndarray]]:
                     """Inner critic loss function: see outer function for parameters."""
 
-                    values = network.critic_network.apply(critic_params, observations)
+                    values = (
+                        network.critic_network.apply(critic_params, observations) * mask
```
Comment (DriesSmit) [inline mava/components/training/losses.py:399]: Again is it necessary to mask here if we mask in the loss function?

**Review by DriesSmit — COMMENTED** [review]
Thanks @ulricharmel :fire: This work looks really nice. I am glad the masking is working and that we are finally getting some results on SMAC :slightly_smiling_face:

File: mava/components/training/losses.py (hunk 10 above)
Comment (ulricharmel) [reply mava/components/training/losses.py:399]: Great catch. This is not necessary because all the operations that follow are element wise. So we just need to mask right at the end. Same as before

File: examples/smac/feedforward/decentralised/run_ippo_smac_death_masking.py (hunk 8 above)
Comment (ulricharmel) [reply examples/smac/feedforward/decentralised/run_ippo_smac_death_masking.py:102]: We probably need to run an experiment to check that but in that case we will be computing truncated GAE that do not benefit from the full episode advantages.  In the onpolicy implementation they do the truncation in shorter sequences for the recurrent network after computing the GAE and the returns.

File: examples/smac/feedforward/decentralised/run_ippo_smac_death_masking.py (hunk 7 above)
Comment (ulricharmel) [reply examples/smac/feedforward/decentralised/run_ippo_smac_death_masking.py:100]: I generally just disable this when I am running experiments locally. I will set it back to True

File: mava/components/training/losses.py (hunk 9 above)
Comment (ulricharmel) [reply mava/components/training/losses.py]: Not necessary

File: examples/smac/feedforward/decentralised/run_ippo_smac_death_masking.py (hunk 8 above)
Comment (DriesSmit) [reply examples/smac/feedforward/decentralised/run_ippo_smac_death_masking.py:100]: Agreed. Maybe in the future we can figure out a way to still use shorter sequences without affecting the GAE.

**Review by DriesSmit — APPROVED** [review]
Thanks @ulricharmel :raised_hands: The PR looks good to me :+1:

**Review by OmaymaMahjoub — APPROVED** [review]
Thanks @ulricharmel :fire:
